### PR TITLE
move type inference for arange into c++

### LINKED ISSUE #17662: torch.arange dtype mismatch with and without jit
```python
import torch

def f():
    return torch.arange(5)

print(f().dtype)
print(torch.jit.script(f)().dtype)
```

prints

```
torch.int64
torch.float32
```

### PR DESCRIPTION
**Pull request by eellison**
Fixes https://github.com/pytorch/pytorch/issues/17662

I'm not sure if `arange` needs to be in python_arg_parser at all, given the schemas in native_functions.yaml. In any case this at least fixes the dytpe mismatch.

In follow up PRs I will try to handle some of the other ops that do type inference at the python level, like randint.

### CODE REVIEW

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: tools/autograd/templates/python_torch_functions.cpp
Comment (zou3519) [inline tools/autograd/templates/python_torch_functions.cpp]: It's worth exploring if we can delete `THPVariable_arange` from this file (templates/python_torch_functions.cpp) and see if we can auto-generate the arange code. From reading the original arange pull request, it looks like the reason why we special-cased it is to support the custom scalarType logic; now that we've pushed the logic to the arange operator itself we shouldn't this manual binding anymore.

**Comment by zou3519** [comment]
cc @yf225 I think this breaks BC for the C++ API. If I understand this correctly,
- before: torch::arange(3) gives a float tensor
- after: torch::arange(3) gives a long tensor.

I'm not 100% sure if this is the case. @eellison it would be good to add a test into the C++ API tests that checks what the dtype of this operation is.

**Comment by yf225** [comment]
Yes it would be great to add a C++ API test for this to understand whether it's BC-breaking. On the other hand, if `torch.arange(3)` in Python API gives a long tensor, C++ API should do the same.

**Review by zou3519 — CHANGES_REQUESTED** [review]
(not a full review) Please don't land yet, I want to take a full pass through this

File: test/cpp/api/functional.cpp
Comment (yf225) [inline test/cpp/api/functional.cpp]: @eellison I marked this PR as BC-breaking, because currently in C++ API, `torch::arange(0, 5).dtype()` prints `float`.

File: test/cpp/api/functional.cpp
Comment (yf225) [inline test/cpp/api/functional.cpp]: This test should live in `test/cpp/api/tensor.cpp`, as `functional.cpp` is for `torch::nn::functional` tests.

**Review by yf225 — DISMISSED** [review]

File: aten/src/ATen/native/TensorFactories.cpp
Comment (zou3519) [inline aten/src/ATen/native/TensorFactories.cpp]: This always does a copy of updated_options. We don't need to do a copy if we're not setting to integral dtype. We can avoid the copy by doing something like:
```
Tensor result;
if (set_to_integral_dtype) {
    result = at::empty({0}, options.dtype(at::ScalarType::Long));
} else {
    result = at::empty({0}, options);
}
return at::arange_out(result, start, end, step);
```

File: test/onnx/expect/TestOperators.test_dyn_arange.expect
Comment (zou3519) [inline test/onnx/expect/TestOperators.test_dyn_arange.expect]: What is going on here? Why does the onnx expect change? Shouldn't onnx follow the python frontend?

File: tools/autograd/templates/python_torch_functions.cpp
Comment (zou3519) [reply tools/autograd/templates/python_torch_functions.cpp]: Okay, I looked into this a little. We can't actually autogenerate `THPVariable_arange` because the generated code uses `r.scalartype(X)`, which "gives the default dtype if r.isNone(X)" (thank you to whoever wrote that comment above this line). 

Given that, our two options are:
1) Fix the python arg parser
2) Accept this change.

I'm inclined to go with option (2) because this change does not make the code any worse than before.

File: tools/autograd/templates/python_torch_functions.cpp
Comment (zou3519) [inline tools/autograd/templates/python_torch_functions.cpp]: Could we add a `optional<ScalarType> PythonArgs::toScalarTypeOptional(int i)` helper method that does this?

File: test/onnx/expect/TestOperators.test_dyn_arange.expect
Comment (eellison) [reply test/onnx/expect/TestOperators.test_dyn_arange.expect]: i dont know, it's possible it has something to do with the dtype changing

File: aten/src/ATen/native/TensorFactories.cpp
Comment (zou3519) [inline aten/src/ATen/native/TensorFactories.cpp]: Should be two spaces (the rest of this file uses two-space indentation)

**Review by zou3519 — DISMISSED** [review]
We should figure out if the onnx change is important or not. I don't want angry onnx folks coming after us.

**Comment by eellison** [comment]
> We should figure out if the onnx change is important or not. I don't want angry onnx folks coming after us.

Sure, though it would be nice if the ONNX tests tested functionality so I know if i've broken something, instead of just "raw data went from `"\001\000\000\000\000\000\000\000"` to `"\000\000\200?"`. JIT removed our expect tests for this reason.

cc: @houseroad

File: tools/autograd/templates/python_torch_functions.cpp
Comment (eellison) [reply tools/autograd/templates/python_torch_functions.cpp]: This actually changes the semantics - `c10::nullopt` for scalar type is different than the default dtype, which is what is happened for `r.scalartype(2);`.

File: test/onnx/expect/TestOperators.test_dyn_arange.expect
Comment (houseroad) [reply test/onnx/expect/TestOperators.test_dyn_arange.expect]: I think it should be the same, could you run `python test/onnx/test_operators.py --accept` to update the expect files?

From the CI, seems they should stay the same.

Comment (eellison) [reply test/onnx/expect/TestOperators.test_dyn_arange.expect]: I updated this because it was failing previously

File: aten/src/ATen/native/TensorFactories.cpp
Comment (gchanan) [inline aten/src/ATen/native/TensorFactories.cpp]: you should probably comment on the boolean case.  I think this implementation is correct (and we never use bool, which matches NumPy), but I'm not 100% sure.

File: test/test_jit.py
Comment (gchanan) [inline test/test_jit.py]: maybe throw in a bool or two here.

**Review by gchanan — COMMENTED** [review]
looks good, I'd just be explicit about what is going on with bools.

I'm also not sure we can't autogenerate this.  If you have a optional ScalarType (like we do here), does the generated code in python_arg_parser and JIT actually fill in the dtype or does it leave it optional?  If it leaves it optional it looks like the code will work (because TensorOptions will do the right thing).

**Comment by zou3519** [comment]
> I'm also not sure we can't autogenerate this. If you have a optional ScalarType (like we do here), does the generated code in python_arg_parser and JIT actually fill in the dtype or does it leave it optional? If it leaves it optional it looks like the code will work (because TensorOptions will do the right thing).

The generated code in python_arg_parser actually fills in the dtype, so we'd have to change the codegen a little to make it work.

**Comment by eellison** [comment]
> looks good, I'd just be explicit about what is going on with bools.
> 
> I'm also not sure we can't autogenerate this. If you have a optional ScalarType (like we do here), does the generated code in python_arg_parser and JIT actually fill in the dtype or does it leave it optional? If it leaves it optional it looks like the code will work (because TensorOptions will do the right thing).



> > I'm also not sure we can't autogenerate this. If you have a optional ScalarType (like we do here), does the generated code in python_arg_parser and JIT actually fill in the dtype or does it leave it optional? If it leaves it optional it looks like the code will work (because TensorOptions will do the right thing).
> 
> The generated code in python_arg_parser actually fills in the dtype, so we'd have to change the codegen a little to make it work.

This would be a nice change to make, but would likely require fixing a number of ops. It's something we should do but not needed for this PR.

File: test/test_jit.py
Comment (eellison) [reply test/test_jit.py]: JIT doesn't allow promotion of bools to scalars, added in test_torch

**Comment by yf225** [comment]
@eellison Wondering what would be the next step for this patch? It would be nice to merge the fix for "if `torch.arange(3)` in Python API gives a long tensor, C++ API should do the same." Thanks!

**Comment by eellison** [comment]
> @eellison Wondering what would be the next step for this patch? It would be nice to merge the fix for "if `torch.arange(3)` in Python API gives a long tensor, C++ API should do the same." Thanks!

I'll try to land it today/tomorrow. Sorry i kept on having to rebase to satisfy the onnx expects and got fatigue. Thx for reminder!

**Comment by yf225** [comment]
@eellison Yes the onnx expects are annoying.. :(  Thanks so much for your help!

**Comment by eellison** [comment]
@yf225 could you keep your eyes out for c++ PRs that use the arange constructor with all integral inputs ? keep on getting test failures after rebasing where a test used `torch::arange` and this broke the test

**Comment by yf225** [comment]
@eellison My apologies for the bad experience - I will be sure to change any new usage of `torch::arange(...)` to `torch::arange(...).to(torch::kFloat)` to unblock this PR from landing.

cc. @pbelevich @anjali411

**Comment by eellison** [comment]
> @eellison My apologies for the bad experience - I will be sure to change any new usage of `torch::arange(...)` to `torch::arange(...).to(torch::kFloat)` to unblock this PR from landing.
> 
> cc. @pbelevich @anjali411

No problem! For context to anyone else i added a comment about not adding any new tests that rely on the previous behavior of `torch::arange`. I then deleted it b/c this only happened once and what i thought was a second time was just my mistake lol

**Comment by eellison** [comment]
More ONNX failures here :'( There isn't 
```
Nov 04 23:37:23 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Nov 04 23:37:23 test/onnx/test_pytorch_onnx_caffe2.py:2025: in test_size
Nov 04 23:37:23     self.run_model_test(SizeModel(), train=False, input=(x,), batch_size=BATCH_SIZE)
Nov 04 23:37:23 test/onnx/test_pytorch_onnx_caffe2.py:201: in run_model_test
Nov 04 23:37:23     do_constant_folding=do_constant_folding)
Nov 04 23:37:23 test/onnx/test_pytorch_onnx_caffe2.py:187: in run_actual_test
Nov 04 23:37:23     keep_initializers_as_inputs=True)
Nov 04 23:37:23 test/onnx/verify.py:372: in verify
Nov 04 23:37:23     prepared = backend.prepare(proto)
Nov 04 23:37:23 ../.local/lib/python3.6/site-packages/caffe2/python/onnx/backend.py:708: in prepare
Nov 04 23:37:23     model = onnx.shape_inference.infer_shapes(model)
Nov 04 23:37:23 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Nov 04 23:37:23 
Nov 04 23:37:23 model = ir_version: 4
Nov 04 23:37:23 producer_name: "pytorch"
Nov 04 23:37:23 producer_version: "1.3"
Nov 04 23:37:23 graph {
Nov 04 23:37:23   node {
Nov 04 23:37:23     input: "0"
Nov 04 23:37:23     output: "1"
Nov 04 23:37:23     op_...ape {
Nov 04 23:37:23           dim {
Nov 04 23:37:23             dim_value: 2
Nov 04 23:37:23           }
Nov 04 23:37:23         }
Nov 04 23:37:23       }
Nov 04 23:37:23     }
Nov 04 23:37:23   }
Nov 04 23:37:23 }
Nov 04 23:37:23 opset_import {
Nov 04 23:37:23   version: 9
Nov 04 23:37:23 }
Nov 04 23:37:23 
Nov 04 23:37:23 
Nov 04 23:37:23     def infer_shapes(model):  # type: (ModelProto) -> ModelProto
Nov 04 23:37:23         if not isinstance(model, ModelProto):
Nov 04 23:37:23             raise ValueError('Shape inference only accepts ModelProto, '
Nov 04 23:37:23                              'incorrect type: {}'.format(type(model)))
Nov 04 23:37:23         model_str = model.SerializeToString()
Nov 04 23:37:23 >       inferred_model_str = C.infer_shapes(model_str)
Nov 04 23:37:23 E       RuntimeError: Inferred elem type differs from existing elem type: (FLOAT) vs (INT64)
Nov 04 23:37:23 
Nov 04 23:37:23 ../.local/lib/python3.6/site-packages/onnx/shape_inference.py:35: RuntimeError
Nov 04 23:37:23 ----------------------------- Captured stderr call -----------------------------
Nov 04 23:37:23 (op_type:Cast, name:): Inferred elem type differs from existing elem type: (FLOAT) vs (INT64)
Nov 04 23:37:23 _____________________ TestCaffe2Backend_opset10.test_size ______________________
Nov 04 23:37:23 
Nov 04 23:37:23 self = <test_pytorch_onnx_caffe2.TestCaffe2Backend_opset10 testMethod=test_size>
Nov 04 23:37:23 
Nov 04 23:37:23     def wrapper(self):
Nov 04 23:37:23         if self.opset_version < min_opset_version:
Nov 04 23:37:23             raise unittest.SkipTest("Skip verify test for unsupported opset_version")
Nov 04 23:37:23 >       return func(self)
Nov 04 23:37:23 
Nov 04 23:37:23 test/onnx/test_pytorch_common.py:54: 
Nov 04 23:37:23 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Nov 04 23:37:23 test/onnx/test_pytorch_onnx_caffe2.py:2025: in test_size
Nov 04 23:37:23     self.run_model_test(SizeModel(), train=False, input=(x,), batch_size=BATCH_SIZE)
Nov 04 23:37:23 test/onnx/test_pytorch_onnx_caffe2.py:204: in run_model_test
Nov 04 23:37:23     use_gpu=use_gpu_, example_outputs=example_outputs)
Nov 04 23:37:23 test/onnx/test_pytorch_onnx_caffe2.py:158: in run_debug_test
Nov 04 23:37:23     caffe2_out = run_embed_params(onnxir, model, input, state_dict, use_gpu)
Nov 04 23:37:23 test/onnx/debug_embed_params.py:36: in run_embed_params
Nov 04 23:37:23     prepared = c2.prepare(model_def, device=device)
Nov 04 23:37:23 ../.local/lib/python3.6/site-packages/caffe2/python/onnx/backend.py:708: in prepare
Nov 04 23:37:23     model = onnx.shape_inference.infer_shapes(model)
Nov 04 23:37:23 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Nov 04 23:37:23 
Nov 04 23:37:23 model = ir_version: 4
Nov 04 23:37:23 producer_name: "pytorch"
Nov 04 23:37:23 producer_version: "1.3"
Nov 04 23:37:23 graph {
Nov 04 23:37:23   node {
Nov 04 23:37:23     input: "0"
Nov 04 23:37:23     output: "1"
Nov 04 23:37:23     op_...pe {
Nov 04 23:37:23           dim {
Nov 04 23:37:23             dim_value: 2
Nov 04 23:37:23           }
Nov 04 23:37:23         }
Nov 04 23:37:23       }
Nov 04 23:37:23     }
Nov 04 23:37:23   }
Nov 04 23:37:23 }
Nov 04 23:37:23 opset_import {
Nov 04 23:37:23   version: 10
Nov 04 23:37:23 }
Nov 04 23:37:23 
Nov 04 23:37:23 
Nov 04 23:37:23     def infer_shapes(model):  # type: (ModelProto) -> ModelProto
Nov 04 23:37:23         if not isinstance(model, ModelProto):
Nov 04 23:37:23             raise ValueError('Shape inference only accepts ModelProto, '
Nov 04 23:37:23                              'incorrect type: {}'.format(type(model)))
Nov 04 23:37:23         model_str = model.SerializeToString()
Nov 04 23:37:23 >       inferred_model_str = C.infer_shapes(model_str)
Nov 04 23:37:23 E       RuntimeError: Inferred elem type differs from existing elem type: (FLOAT) vs (INT64)
```

It looks like error is because onnx shape inference (which is outside of this repo) is relying on the previous behavior of `arange`. I think the move here is to disable the test and land and when onnx updates its shape inference we can re-enable the test. Thoughts ? @yf225 @houseroad

**Comment by houseroad** [comment]
This sounds suspicious. @BowenBao, @lara-hdr, and @spandantiwari, could one of you guys look into the issue and confirm the root cause?

**Comment by eellison** [comment]
@houseroad there’s no shape analysis registered for arange within this code base, and the error message is that there is a dtype mismatch. “ Inferred elem type differs from existing elem type: (FLOAT) vs (INT64) “. 

What do you mean by “this sounds suspicious”?

**Comment by BowenBao** [comment]
```python
    @skipIfUnsupportedMinOpsetVersion(9)
    def test_size(self):
        class SizeModel(torch.nn.Module):
            def forward(self, input):
                return torch.arange(input.size(0)), torch.arange(input.size(-1))

        x = torch.randn(5, 3, 2)
        self.run_model_test(SizeModel(), train=False, input=(x,), batch_size=BATCH_SIZE)
```
This is the failing test. The mismatch error seems strange as the arange here only takes 1 input, I wonder where FLOAT is from.

For current master, the test produces the following PTIR graph, before being converted to ONNX. There's nothing with type FLOAT except for the input which is irrelevant.
```
graph(%0 : Float(5, 3, 2)):
  %22 : Long() = prim::Constant[value={0}](), scope: SizeModel
  %4 : Long() = aten::size(%0, %22), scope: SizeModel # test_pytorch_onnx_caffe2.py:2022:0
  %23 : Long() = prim::Constant[value={4}](), scope: SizeModel
  %24 : Long() = prim::Constant[value={0}](), scope: SizeModel
  %9 : Device = prim::Constant[value="cpu"](), scope: SizeModel # test_pytorch_onnx_caffe2.py:2022:0
  %25 : Long() = prim::Constant[value={0}](), scope: SizeModel
  %11 : Long(5) = aten::arange(%4, %23, %24, %9, %25), scope: SizeModel # test_pytorch_onnx_caffe2.py:2022:0
  %26 : Long() = prim::Constant[value={-1}](), scope: SizeModel
  %13 : Long() = aten::size(%0, %26), scope: SizeModel # test_pytorch_onnx_caffe2.py:2022:0
  %27 : Long() = prim::Constant[value={4}](), scope: SizeModel
  %28 : Long() = prim::Constant[value={0}](), scope: SizeModel
  %18 : Device = prim::Constant[value="cpu"](), scope: SizeModel # test_pytorch_onnx_caffe2.py:2022:0
  %29 : Long() = prim::Constant[value={0}](), scope: SizeModel
  %20 : Long(2) = aten::arange(%13, %27, %28, %18, %29), scope: SizeModel # test_pytorch_onnx_caffe2.py:2022:0
  return (%11, %20)
```

Is the PTIR graph different after this PR?(Things like default parameters/arguments).

**Comment by eellison** [comment]
> ```python
>     @skipIfUnsupportedMinOpsetVersion(9)
>     def test_size(self):
>         class SizeModel(torch.nn.Module):
>             def forward(self, input):
>                 return torch.arange(input.size(0)), torch.arange(input.size(-1))
> 
>         x = torch.randn(5, 3, 2)
>         self.run_model_test(SizeModel(), train=False, input=(x,), batch_size=BATCH_SIZE)
> ```
> 
> This is the failing test. The mismatch error seems strange as the arange here only takes 1 input, I wonder where FLOAT is from.
> 
> For current master, the test produces the following PTIR graph, before being converted to ONNX. There's nothing with type FLOAT except for the input which is irrelevant.
> 
> ```
> graph(%0 : Float(5, 3, 2)):
>   %22 : Long() = prim::Constant[value={0}](), scope: SizeModel
>   %4 : Long() = aten::size(%0, %22), scope: SizeModel # test_pytorch_onnx_caffe2.py:2022:0
>   %23 : Long() = prim::Constant[value={4}](), scope: SizeModel
>   %24 : Long() = prim::Constant[value={0}](), scope: SizeModel
>   %9 : Device = prim::Constant[value="cpu"](), scope: SizeModel # test_pytorch_onnx_caffe2.py:2022:0
>   %25 : Long() = prim::Constant[value={0}](), scope: SizeModel
>   %11 : Long(5) = aten::arange(%4, %23, %24, %9, %25), scope: SizeModel # test_pytorch_onnx_caffe2.py:2022:0
>   %26 : Long() = prim::Constant[value={-1}](), scope: SizeModel
>   %13 : Long() = aten::size(%0, %26), scope: SizeModel # test_pytorch_onnx_caffe2.py:2022:0
>   %27 : Long() = prim::Constant[value={4}](), scope: SizeModel
>   %28 : Long() = prim::Constant[value={0}](), scope: SizeModel
>   %18 : Device = prim::Constant[value="cpu"](), scope: SizeModel # test_pytorch_onnx_caffe2.py:2022:0
>   %29 : Long() = prim::Constant[value={0}](), scope: SizeModel
>   %20 : Long(2) = aten::arange(%13, %27, %28, %18, %29), scope: SizeModel # test_pytorch_onnx_caffe2.py:2022:0
>   return (%11, %20)
> ```
> 
> Is the PTIR graph different after this PR?(Things like default parameters/arguments).
```
graph(%self : ClassType<Module>,
      %input : Float(5, 3, 2)):
  %2 : int = prim::Constant[value=0](), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %3 : int = aten::size(%input, %2), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %4 : Long() = prim::NumToTensor(%3), scope: SizeModel
  %5 : Scalar = prim::ImplicitTensorToNum(%4), scope: SizeModel
  %6 : int = prim::Constant[value=6](), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %7 : int = prim::Constant[value=0](), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %8 : Device = prim::Constant[value="cpu"](), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %9 : bool = prim::Constant[value=0](), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %10 : Long(5) = aten::arange(%5, %6, %7, %8, %9), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %11 : int = prim::Constant[value=-1](), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %12 : int = aten::size(%input, %11), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %13 : Long() = prim::NumToTensor(%12), scope: SizeModel
  %14 : Scalar = prim::ImplicitTensorToNum(%13), scope: SizeModel
  %15 : int = prim::Constant[value=6](), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %16 : int = prim::Constant[value=0](), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %17 : Device = prim::Constant[value="cpu"](), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %18 : bool = prim::Constant[value=0](), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %19 : Long(2) = aten::arange(%14, %15, %16, %17, %18), scope: SizeModel # test/onnx/test_pytorch_onnx_caffe2.py:2011:0
  %20 : (Long(5), Long(2)) = prim::TupleConstruct(%10, %19)
  return (%20)
```

**Comment by BowenBao** [comment]
@eellison Thanks for providing the graph. The difference is the dtype parameter in arange.
```
  %23 : Long() = prim::Constant[value={4}]()
...
  %11 : Long(5) = aten::arange(%4, %23, %24, %9, %25)
```
vs
```
  %6 : int = prim::Constant[value=6]()
...
  %10 : Long(5) = aten::arange(%5, %6, %7, %8, %9)
```
Previously `torch.int64(4)` vs now `torch.float(6)`. Setting the unprovided dtype to float doesn't seem correct, given that the arange returns Long type.

File: test/cpp/api/functional.cpp
Comment (houseroad) [inline test/cpp/api/functional.cpp]: why do we change this test?

**Review by houseroad — COMMENTED** [review]
I think after the changes in this PR, default dtype is set to float, however, the result is still long.

File: test/cpp/api/functional.cpp
Comment (eellison) [reply test/cpp/api/functional.cpp]: to preserve the behavior of the test, since that initializer went from returning a float tensor -> int tensor

**Comment by BowenBao** [comment]
So far here's the discovery. The issue is after the PR, traced graph sets default dtype to float. 

The tracer is not able to call type inference in aten. After moving type inference away in the parser, It will set has_dtype to false for options here
https://github.com/pytorch/pytorch/blob/0fe02515830d261bf51390e00b00ac7ba0632670/tools/autograd/templates/python_torch_functions.cpp#L115

Then the tracer will try to add inputs for options in `VariableTypeEverything.cpp`
```c++
    jit::tracer::addInputs(node, "options", options);
```

which calls 
https://github.com/pytorch/pytorch/blob/f93a6e54b9ab2077103f18aee735e26f3c029eb7/torch/csrc/jit/tracer.cpp#L545

that leads to 

https://github.com/pytorch/pytorch/blob/f93a6e54b9ab2077103f18aee735e26f3c029eb7/c10/core/TensorOptions.h#L269

end up as float.

So the tracer still needs type inference up front in order to preserve previous behavior.

**Comment by eellison** [comment]
> So far here's the discovery. The issue is after the PR, traced graph sets default dtype to float.
> 
> The tracer is not able to call type inference in aten. After moving type inference away in the parser, It will set has_dtype to false for options here
> https://github.com/pytorch/pytorch/blob/0fe02515830d261bf51390e00b00ac7ba0632670/tools/autograd/templates/python_torch_functions.cpp#L115
> 
> Then the tracer will try to add inputs for options in `variable_factories.h`
> 
> ```c++
>     jit::tracer::addInputs(node, "options", options);
> ```
> 
> which calls
> https://github.com/pytorch/pytorch/blob/f93a6e54b9ab2077103f18aee735e26f3c029eb7/torch/csrc/jit/tracer.cpp#L545
> 
> that leads to
> 
> https://github.com/pytorch/pytorch/blob/f93a6e54b9ab2077103f18aee735e26f3c029eb7/c10/core/TensorOptions.h#L269
> 
> end up as float.
> 
> So the tracer still needs type inference up front in order to preserve previous behavior.

We just need to update the onnx export to work with this patch. putting the logic up front still retains the previous problem, which is that `arange` has different semantics in python from JIT & C++ api.

**Comment by BowenBao** [comment]
After this patch the traced graph is different. I don't think that should be expected. Fixing JIT & C++ should not affect tracing. If you keep the shape inference logic in both places, things will work out.

**Comment by eellison** [comment]
> After this patch the traced graph is different. I don't think that should be expected. Fixing JIT & 
> C++ should not affect tracing. If you keep the shape inference logic in both places, things will work out.

That is expected. The tracer captures the values as they are called into the op. 
The signature of arange has an optional-dtype:
`arange(Scalar end, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)`

The export for onnx doesn't match this signature because it doesn't have an optional dtype. 

https://github.com/pytorch/pytorch/blob/ebc216a0765d85f345f9a5cd1dfd2ec360de3a52/torch/onnx/symbolic_opset11.py#L287

If onnx is to export this op, it should be able to handle the pytorch signature. I don't think we should duplicate code in both the pytorch & c++ layer, that's an anti-pattern.

**Comment by BowenBao** [comment]
Please check the above example. Running traced module now produces different type result.

**Comment by eellison** [comment]
> Please check the above example. Running traced module now produces different type result.

Thank you for pointing that out. This is because of legacy behavior in TensorOptions that returns the default dtype if it doesn't have one. See: https://github.com/pytorch/pytorch/blob/master/c10/core/TensorOptions.h#L280

**Comment by eellison** [comment]
> Please check the above example. Running traced module now produces different type result.

Just updated the tracer to make it more correct, and included the example you gave as a test.

**Comment by eellison** [comment]
@BowenBao just updated tracer to make it more correct and understand optional dtypes, and included the example you gave as a test.

**Comment by BowenBao** [comment]
Thank you @eellison, this fixes the tracer for arange. But i'm worried this also affect export of other ops. It is bc-breaking for ops that rely on tracer setting the default type for them, and we don't know how many such ops are out there. For arange I have created a PR #29399 working with the new traced graph.

**Comment by eellison** [comment]
> Thank you @eellison, this fixes the tracer for arange. But i'm worried this also affect export of other ops. It is bc-breaking for ops that rely on tracer setting the default type for them, and we don't know how many such ops are out there. For arange I have created a PR #29399 working with the new traced graph.

Cool! I think it's just `arange` currently, because of the above mentioned behavior where if `TensorOptions` doesn't have a `dtype` it will just return current default dtype. I think all the tests were passing with this change.

**Comment by eellison** [comment]
@BowenBao looks like this causes size to fail as well. I'm not sure what the exact path forward is here. We can either:
- duplicate dtype inference in python layer for now, and then land the changes to the tracer & arange/size onnx changes in upcoming diffs
- try to get the arange & size onnx changes in this PR

**Comment by BowenBao** [comment]
@eellison I'd prefer the first option. With two PRs, we can land the jit fix for arange, and if things do break with the tracer change we can revert and fix things without affecting the jit fix. @houseroad @spandantiwari

**Comment by eellison** [comment]
> @eellison I'd prefer the first option. With two PRs, we can land the jit fix for arange, and if things do break with the tracer change we can revert and fix things without affecting the jit fix. @houseroad @spandantiwari

Sure, that's good with me. Let's just try to get the other todos fixed in a reasonable time frame. It doesn't look like it will be too much work, just including the tracer changes i have right now and then updating `arange` / `size` onnx export.

**Comment by BowenBao** [comment]
@eellison - Sounds great, I have updated my PR https://github.com/pytorch/pytorch/pull/29399 with tracer change + dup inference removal. Once this is merged I will go ahead and rebase with master.

**Comment by eellison** [comment]
@BowenBao i'm still getting ONNX errors even with dtype inference in front-end:
`test_arange_end_notype`
`test_arange_end_notype`

**Comment by BowenBao** [comment]
> @BowenBao i'm still getting ONNX errors even with dtype inference in front-end:
> `test_arange_end_notype`
> `test_arange_end_notype`

Thanks for catching that. These are testing scripted arange. So the previous onnx export for that part actually tried to match the wrong type ... Let's include the onnx export fix in this PR then.

https://github.com/pytorch/pytorch/pull/29399/files#diff-4e5d6309f038e27dc5caefdb078728ac

**Comment by eellison** [comment]
> > @BowenBao i'm still getting ONNX errors even with dtype inference in front-end:
> > `test_arange_end_notype`
> > `test_arange_end_notype`
> 
> Thanks for catching that. These are testing scripted arange. So the previous onnx export for that part actually tried to match the wrong type ... Let's include the onnx export fix in this PR then.
> 
> https://github.com/pytorch/pytorch/pull/29399/files#diff-4e5d6309f038e27dc5caefdb078728ac

I think the export for `size` needs to be updated too.

**Comment by BowenBao** [comment]
@eellison  that test is also calling arange, which is the part that was actually failing. The patch for arange should be enough.

**Pull request closed without merging**
